src/init/bitcoind.cpp

```diff
@@ -18,6 +22,13 @@ class BitcoindInit : public interfaces::Init
         m_node.args = &gArgs;
         m_node.init = this;
     }
+    std::unique_ptr<interfaces::Node> makeNode() override { return interfaces::MakeNode(m_node); }
+    std::unique_ptr<interfaces::Chain> makeChain() override { return interfaces::MakeChain(m_node); }
+    std::unique_ptr<interfaces::WalletClient> makeWalletClient(interfaces::Chain& chain) override
+    {
+        return MakeWalletClient(chain, *Assert(m_node.args));
+    }
+    std::unique_ptr<interfaces::Echo> makeEcho() override { return interfaces::MakeEcho(); }
```

Comment:
What's the rational of having an interfaces maker such as `makeEcho` supported on `BitcoindInit`, `bitcoind` program should never be executed as a spawned process/`IpcServer` ?

One option is to tread this as a throw returned to the client, as `MakeWalletClient` is providing, though it means anytime we add a new derived Init interface, we need to add throw stub in all other programs non-supporting the interface functionality ?

Currently, if the interface maker isn't implemented-side, what I can observe on one of my local branch is silent failure of the IPC server, followed by a socket disconnection/`EventLoop` exit.

```
2021-07-08T17:07:01Z {bitcoin-altnet-26997/bitcoin-altnet-26997} IPC server recv request  #1 Init.construct$Params (threadMap = <external capability>)                                                                                                                                   
2021-07-08T17:07:01Z
2021-07-08T17:07:01Z {bitcoin-altnet-26997/bitcoin-altnet-26997} IPC server send response #1 Init.construct$Results (threadMap = <external capability>)                                                                                                                                  
2021-07-08T17:07:01Z
2021-07-08T17:07:01Z {bitcoin-altnet-26997/bitcoin-altnet-26997} IPC server recv request  #2 Init.makeEcho$Params (context = (thread = <external capability>, callbackThread = <external capability>))                                                                                   
2021-07-08T17:07:01Z
2021-07-08T17:07:01Z {bitcoin-altnet-26997/bitcoin-altnet-26997} IPC server post request  #2 {bitcoin-altnet-26997/bitcoin-altnet-26999 (from bitcoin-node-26897/b-httpworker.0-26916)}                                                                                                  
2021-07-08T17:07:01Z
2021-07-08T17:07:01Z {bitcoin-altnet-26997/bitcoin-altnet-26997} IPC server send response #2 Init.makeEcho$Results ()
2021-07-08T17:07:01Z
2021-07-08T17:07:01Z {bitcoin-altnet-26997/bitcoin-altnet-26997} IPC server: socket disconnected.
2021-07-08T17:07:01Z
2021-07-08T17:07:01Z {bitcoin-altnet-26997/bitcoin-altnet-26997} IPC server destroyN2mp11ProxyServerIN3ipc5capnp8messages4InitEEE
2021-07-08T17:07:01Z
2021-07-08T17:07:01Z {bitcoin-altnet-26997/bitcoin-altnet-26997} EventLoop::loop done, cancelling event listeners.
2021-07-08T17:07:01Z
2021-07-08T17:07:01Z {bitcoin-altnet-26997/bitcoin-altnet-26997} EventLoop::loop bye.
```

Reply:
re: https://github.com/bitcoin/bitcoin/pull/22219#discussion_r666378196

Thanks for experimenting with this!

> What's the rational of having an interfaces maker such as `makeEcho` supported on `BitcoindInit`, `bitcoind` program should never be executed as a spawned process/`IpcServer` ?

Just like all interfaces in the `src/interfaces/` directory, the `interfaces::Init` interface can be used remotely or it can be used locally. `bitcoind` provides the `interfaces::Chain` interface and it uses the `interface::WalletClient` interface even though it does not support IPC at all. `bitcoin-qt` uses the `interfaces::Node` and `interfaces::Wallet` interfaces even though it does not support IPC at all.

The `interfaces::Init` interface is just the starting interface. When a bitcoin process is created, whether or not uses IPC it can create an instance of the `interfaces::Init` interface and use it to get access to other interfaces supported by the process. If the current process like `bitcoind` and `bitcoin-qt` and `bitcoin-node` has node code linked in, then `makeNode()` and `makeChain()` methods will return non-null. If the current process like `bitcoind` and `bitcoin-qt` and `bitcoin-wallet` has wallet code linked in, then `makeWalletClient` will return non-null. If the current process like `bitcoin-gui` has none of these things linked in then all of these methods will return null.

The point of the Init interface is to allow the same bitcoin init code to be linked into different binaries and work without having to use preprocessor conditions or other build tricks. `make*` methods just return non-null if the relevant interface implementation is linked in and null if trying to instantiate the interface would cause link errors.

Everything described above is happens with or without IPC. IPC just adds the option of calling other `interface::Init` instances that aren't in the current process. You can spawn a new process and call its Init methods. Or you can connect to an existing process through a socket can call its Init methods. Or you can create a socket and let other processes call your process's Init methods.

> One option is to tread this as a throw returned to the client, as `MakeWalletClient` is providing, though it means anytime we add a new derived Init interface, we need to add throw stub in all other programs non-supporting the interface functionality ?

I'm not clear what this would be trying to do. Is it is related to the bug described below? Returning does null seems more ergonomic than throwing an exception. libmultiprocess does support throwing & catching exceptions remotely, but it requires an extra annotation in capnp interface definitions.

> Currently, if the interface maker isn't implemented-side, what I can observe on one of my local branch is silent failure of the IPC server, followed by a socket disconnection/`EventLoop` exit.

This isn't expected and sounds like a bug. Can you post a branch or commit hash with the code? I think there should never be much reason to call a `make` method in a remote process if you know it will return null. But if it does return null you should just see a null return value forwarded. It shouldn't cause the remote process to exit.

The log is a little strange, and I'm not sure how to interpret it without looking at the code (it might also help to see the client side log). It looks like the server is successfully returning null from a remote `makeEcho` request. But then the socket is disconnected and it exits. Maybe the socket is closing because the client is segfaulting?

Would be great to post details in new issue (https://github.com/chaincodelabs/libmultiprocess/issues/new) since that issue tracker is a convenient place to collect feedback and provide support for these things.

Reply:
Thanks to recall the design of `interfaces::Init`, this is matching the rough mental model I've in mind!

> If the current process like bitcoin-gui has none of these things linked in then all of these methods will return null.

This is where I think the interfaces error handling could be improved. Instead of returning null if the current process doesn't have linked code, what do you think about throwing back an exception from a stub ?

AFAIU, this error handling is what the newly introduced `MakeWalletClient` in `src/dummywallet.cpp` (https://github.com/bitcoin/bitcoin/pull/22219/files#diff-ccc5d781a55644499897b7329c1544939c0751975d687abbda086d332db21fe7R68) is achieving. I was proposing to extend it to `makeEcho` on `bitcoind` build. Like https://github.com/ariard/bitcoin/commit/1dee88ad1bf4434bc495b77566c7e617245095a5.

But AFAICT, you think it's more ergonomic to return null rather than throwing an exception where I would prefer more explicit, verbose failure handling ? That said, I'm good with returning null, just a small ergonomic preference here.

-------

> Maybe the socket is closing because the client is segfaulting?

Yes I'm deliberately calling a method on the returned, null pointer to demonstrate the failure mode, sorry I should have precise that. Hopefully, it's illustrating that throwing on null `make` method might be better to recover from logical bugs introduced by programmers mistakenly calling a `make`.

> Would be great to post details in new issue (https://github.com/chaincodelabs/libmultiprocess/issues/new) since that issue tracker is a convenient place to collect feedback and provide support for these things.

Still cleaning the code on this branch, will post it for sure for feedback but so far I'm able to chain processes (`bitcoind` --> `altnet_1` --> `altnet_2`) as intended :)

Reply:
re: https://github.com/bitcoin/bitcoin/pull/22219#discussion_r668024917

> But AFAICT, you think it's more ergonomic to return null rather than throwing an exception where I would prefer more explicit, verbose failure handling ? That said, I'm good with returning null, just a small ergonomic preference here.

I think it's just a misunderstanding. I like exceptions and think it's often good to throw when a condition happens that is definitely an error. But if an object can't be created locally so it has to be created remotely, that may be an error, or it may be fine. The `interfaces::Init` make methods usually have no way of knowing what their callers are expecting. So I think all they should do is return the objects if they can make them or return null if they can't make them.

If you look at the `makeNode` call in the GUI code:

https://github.com/ryanofsky/bitcoin/blob/pr/ipc.157/src/qt/bitcoin.cpp#L284-L285

Or the `makeWalletClient` code in the node code:

https://github.com/ryanofsky/bitcoin/blob/pr/ipc.157/src/wallet/init.cpp#L134-L135

You can see that it's not an error, but an expected condition for either of these methods to return null. If the node or wallet client object can't be created locally (the respective make methods return null in `bitcoin-gui` and `bitcoin-node` binaries and non-null in `bitcoin-qt` and `bitcoind` binaries), the null-handling code simply calls `spawnProcess` and creates the object remotely.

If you look at `makeEcho` code in the `echoipc` test:

https://github.com/ryanofsky/bitcoin/blob/pr/ipc.157/src/rpc/misc.cpp#L685

Then it is doing the opposite. For testing purposes it preferring to create the object remotely, but will fall back and create it locally if IPC is not available. This is a test design decision that I think should be local to the `echoipc` function, and not something which five different `Init` classes should be assuming. I think the make methods should all just make the object if they can or return null if they can't and leave callers to handle the result.

> Still cleaning the code on this branch, will post it for sure for feedback but so far I'm able to chain processes (`bitcoind` --> `altnet_1` --> `altnet_2`) as intended :)

Glad to hear it!